### PR TITLE
Fix crate version json parsing

### PR DESCRIPTION
**Pull request by avrong**
Bug introduced in https://github.com/intellij-rust/intellij-rust/pull/6591

As it turns out, both `Gson` and `Jackson` can't work with inside-function classes.

Citation from https://github.com/FasterXML/jackson-module-kotlin/issues/135:
> It sounds like inner class you use would be non-static (as per exception). Such classes can not be deserialized since there is no access to implicit this reference that would be needed, at least on regular Java. I don't know if Kotlin allows some way make inner class static from JVM perspective; if so, that would make things work fine.

`Gson` is returning `null` in this case without any hint of what could go wrong.

This PR fixes the parsing logic by moving `ParsedVersion` out from the function and uses Jackson as it gives more understandable messages and has better Kotlin support. We might also try to use `kotlinx.serialization` instead, but it may need to have some modifications in Gradle plugins.

In my environment, it now works with no one warning about failed-to-parse versions.

### CODE REVIEW

**Comment by Kobzol** [comment]
Heh, it seemed like an innocent change :) Good catch, thanks.

File: toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexServiceImpl.kt [hunk 1]
```diff
@@ -314,15 +316,17 @@ private object CrateExternalizer : DataExternalizer<CargoRegistryCrate> {
...
+data class ParsedVersion(
+    @JsonProperty("name") val name: String,
+    @JsonProperty("vers") val vers: String,
+    @JsonProperty("yanked") val yanked: Boolean,
+    @JsonProperty("features") val features: HashMap<String, List<String>>
+)
 
-    val parsedVersion = Gson().fromJson(json, ParsedVersion::class.java)
+private fun crateFromJson(json: String): CargoRegistryCrateVersion {
+    val parsedVersion = ObjectMapper()
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexServiceImpl.kt:327]: Consider using Kotlin module this way:
```Kotlin
    val parsedVersion = JsonMapper()
        .registerKotlinModule()
        .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
        .readValue<ParsedVersion>(json)
```
See `org.rustPerformanceTests.CustomRealProjectAnalysisTest.Companion#data` for example.

File: toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexServiceImpl.kt [hunk 2]
```diff
@@ -314,15 +316,17 @@ private object CrateExternalizer : DataExternalizer<CargoRegistryCrate> {
...
 
-private fun crateFromJson(json: String): CargoRegistryCrateVersion {
-    data class ParsedVersion(
-        val name: String,
-        val vers: String,
-        val yanked: Boolean,
-        val features: HashMap<String, List<String>>
-    )
+data class ParsedVersion(
+    @JsonProperty("name") val name: String,
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexServiceImpl.kt:320]: Looks like these annotations are redundant if the Kotlin module is used (see the comment below)

**Review by ortem — CHANGES_REQUESTED** [review]

**Comment by ortem** [comment]
BTW I haven't managed to reproduce the bug, but I faced a different problem:
bug

@avrong Is this related to the original issue or JSON parsing?

**Comment by ortem** [comment]
> Heh, it seemed like an innocent change :)

@Kobzol Yep, and it's a funny thing that I suggested this breaking change in the review actually, sorry for that 😄

**Comment by avrong** [comment]
@ortem That's interesting. I fixed the problem that the log warnings contained, but now see it is different from the one explained in the issue. Even with the version info parsing problems, there should be no problems with crate names by themselves.

The original problem might be related to @Kobzol case, as he mentioned. But I haven’t managed to reproduce it too.

In your case, version parsing seems to be the problem. You can invalidate caches to force index load from scratch

**Review by ortem — APPROVED** [review]
LGTM. Just to sum up: this **does not** fix https://github.com/intellij-rust/intellij-rust/issues/6812, but anyway, this change is necessary to parse JSON properly. As for https://github.com/intellij-rust/intellij-rust/issues/6812, it should be investigated and fixed in a separate PR.

bors r+

**Comment by avrong** [comment]
bors r=ortem

**Pull request merged**